File: XUCore/additions/NSViewAdditions.m
// 
// NSViewAdditions.m
// 
// Created by Charlie Monroe
// 
// Copyright (c) 2010-13 Fuel Collective, LLC. All rights reserved.
// 


#import "NSViewAdditions.h"


@implementation NSView (NSViewAdditions)
-(NSRect)screenCoordinatesOfFrame:(NSRect)frame{
	NSRect rect = [self convertRect:frame toView:nil];
	NSRect windowFrame = [[self window] frame];
	rect.origin.x += windowFrame.origin.x;
	rect.origin.y += windowFrame.origin.y;
	return rect;
}
-(NSPoint)screenCoordinatesOfPoint:(NSPoint)point{
	point = [self convertPoint:point toView:nil];
	NSRect windowFrame = [[self window] frame];
	point.x += windowFrame.origin.x;
	point.y += windowFrame.origin.y;
	return point;
}
-(NSRect)screenCoordinatesOfSelf{
	return [self screenCoordinatesOfFrame:[self bounds]];
}
-(void)setDeepEnabled:(BOOL)flag{
	for (NSView *view in [self subviews]){
		if ([view respondsToSelector:@selector(setEnabled:)]){
			[(NSControl*)view setEnabled:flag];
		}else{
			[view setDeepEnabled:flag];
		}
	}
}
@end

File: XUCore/additions/NSViewAdditions.swift
//
//  NSViewAdditions.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/8/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Foundation

public extension NSView {
	
	private func _enclosingView<T: NSView>() -> T? {
		var view: NSView? = self
		while view != nil {
			if let targetView = view as? T {
				return targetView
			}
			
			view = view!.superview
		}
		return nil
	}
	
	/// Returns enclosing table row view or nil.
	public var enclosingTableRowView: NSTableRowView? {
		return self._enclosingView()
	}
	
	/// Returns enclosing table view or nil.
	public var enclosingTableView: NSTableView? {
		return self._enclosingView()
	}
	
	/// Sets enabled on subviews.
	public func setDeepEnabled(_ flag: Bool) {
		for view in self.subviews {
			if let control = view as? NSControl {
				control.isEnabled = flag
			}else{
				view.setDeepEnabled(flag)
			}
		}
	}
		
	public var screenCoordinates: CGRect {
		return self.screenCoordinates(ofRect: self.bounds)
	}
	
	public func screenCoordinates(ofRect frame: CGRect) -> CGRect {
		if self.window == nil {
			return CGRect()
		}
		
		var rect = self.convert(frame, to: nil)
		let windowFrame = self.window!.frame
		rect.origin.x += windowFrame.minX
		rect.origin.y += windowFrame.minY
		return rect
	}
	
	public func screenCoordinates(ofPoint point: CGPoint) -> CGPoint {
		if self.window == nil {
			return CGPoint()
		}
		
		var localPoint = self.convert(point, to: nil)
		let windowFrame = self.window!.frame
		localPoint.x += windowFrame.minX
		localPoint.y += windowFrame.minY
		return localPoint
	}
	
	
}
